Add Base.each_page

krallin opened the pull request:
Unlike Base.all, this allowss consumer to exit the block as soon as they
have collected everything they need. It's a lower-level method, but it
can be useful when the collection is expected to be lengthy (e.g.
operations, or backups), but only a small initial subset is needed.

---

cc @fancyremarker . 

FWIW, I'm not a huge fan of my implementation here, which from a consumer perspective isn't ideally usable (the consumer needs to call `Aptible::Api::Base.each_page(href: ..., token: ...)`), but I think it's the best we can do considering how the API client is currently architectured (that is, considering that related lists are actually naïve *lists*, not an object that could implement those methods).

krallin commented:
A couple extra comments:

- It goes without saying that this is primarily for the `backup:list` command in the CLI, although we could (should) also use it in Sweetness when checking for operation conflicts.
- I did refactor the tests (they were testing the implementation more than the outcome), but I confirmed that the old implementation passes the new tests just the same.

krallin commented:
@fancyremarker added support for `each_RELATION` as we discussed, which lets you iterate over objects one by one and handles paging behind the scenes (and stops querying pages as soon as you stop handling results of course).

It doesn't do any caching (unlike `RELATIONs`), which I think is probably a good thing: it might be a little unpredictable otherwise.

krallin commented:
Hmmm.. The test fails on JRuby, which is a little unfortunate. It definitely feels like something that should be allowed in Ruby though (and it works on MRI)... I'll try and see if we can refactor this to make it work with Ruby, but I'm not sure :(

fancyremarker commented:
@krallin: It's fine to drop JRuby from the build matrix. That's a relic of a time when we wanted to have the option to switch our APIs to use JRuby, but we've since dropped JRuby from the build matrices of those projects (see e.g. aptible/api.aptible.com#253)

krallin commented:
@fancyremarker FWIW, I think it's nice trying to remain compatible with it if possible, and as it turns out this issue is fixed in newer releases of JRuby. I've updated the testing matrix to use a newer one, but a side-effect is that it takes a little while (2 minutes) to install.

I've also added a bunch of other Rubies for the sake of good coverage against versions, let me know what you think there as well.

Cheers,

fancyremarker commented:
This looks great! Merging.
